File: Go/delete_node.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
    if root == nil {return nil}
    if key < root.Val { 
        root.Left = deleteNode(root.Left, key)
    } else if key > root.Val {
        root.Right = deleteNode(root.Right, key)
    } else {
        if root.Left == nil {return root.Right}
        if root.Right == nil {return root.Left}
        leftMax := root.Left
        for leftMax.Right != nil {leftMax = leftMax.Right}
        leftMax.Right = root.Right
        root = root.Left
    }
    return root
}

File: Java/DeleteNode.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if(root == null) return root;
        
        if(key < root.val) {
            root.left = deleteNode(root.left, key);
        } else if(key > root.val) {
            root.right = deleteNode(root.right, key);
        } else {
            if(root.left == null) return root.right;
            if(root.right == null) return root.left;

            TreeNode leftMost = root.left;
            while(leftMost.right != null) leftMost = leftMost.right;
            leftMost.right = root.right;
            root = root.left;
        }

        return root;
    }
}